src/azure-cli-testsdk/azure/cli/testsdk/base.py

```diff
@@ -136,13 +135,38 @@ def _merge_lists(base, patches):
             recording_name=recording_name
         )
 
+    def setUp(self):
+        if self.random_config_dir:
+            from unittest.mock import patch
+            from azure.cli.core.util import random_string, rmtree_with_retry
+            from azure.cli.core._config import GLOBAL_CONFIG_DIR
+            from knack.util import ensure_dir
+
+            config_dir = os.path.join(GLOBAL_CONFIG_DIR, 'dummy_cli_config_dir', random_string())
+            self.addCleanup(rmtree_with_retry, config_dir)
+            logger.warning('Using random config dir: %s', config_dir)
+            # Knack prioritizes the AZURE_CONFIG_DIR env over the config_dir param, and other functions may call
+            # get_config_dir directly. We need to set the env to make sure the config_dir is used.
+            env_patch = patch.dict(os.environ, {'AZURE_CONFIG_DIR': config_dir})
+            env_patch.start()
+            self.addCleanup(env_patch.stop)
```

Comment:
The official doc of `unittest.mock` also prefers `addCleanup` over `tearDown`: https://docs.python.org/3/library/unittest.mock.html#patch-methods-start-and-stop

![image](https://github.com/Azure/azure-cli/assets/4003950/b344a1de-6eff-4b1a-a399-14a278b8c053)